File: firebase/dart_functions/lib/github/workflow_parser.dart
library;

bool matchesTrigger(
  Map<String, dynamic> parsed,
  String triggerType,
  String? triggerBranch,
) {
  final on = parsed['on'];
  if (on == null) return false;

  final yamlTriggerKey = triggerType == 'pullRequest'
      ? 'pull_request'
      : triggerType;

  if (on is String) {
    return on == yamlTriggerKey;
  }

  if (on is List) {
    return on.contains(yamlTriggerKey);
  }

  if (on is Map) {
    final triggerConfig = on[yamlTriggerKey];
    if (triggerConfig == null && !on.containsKey(yamlTriggerKey)) return false;
    if (triggerConfig == null) return true; // key exists with null value

    if (triggerConfig is Map && triggerConfig['branches'] != null) {
      final branches = triggerConfig['branches'];
      final branchList = branches is List
          ? branches.cast<String>()
          : [branches.toString()];
      if (triggerBranch != null && !branchList.contains(triggerBranch)) {
        return false;
      }
    }

    return true;
  }

  return false;
}

class JobInfo {
  final String jobKey;
  final List<String> needs;
  final String? runsOn;
  final List<JobStep> steps;

  JobInfo({
    required this.jobKey,
    required this.needs,
    required this.runsOn,
    required this.steps,
  });
}

class JobStep {
  final String name;
  final String? run;
  final String? uses;
  final Map<String, String>? withParams;

  JobStep({required this.name, this.run, this.uses, this.withParams});
}

List<JobInfo> extractJobs(Map<String, dynamic> parsed) {
  final jobs = parsed['jobs'];
  if (jobs == null || jobs is! Map) return [];

  final jobInfos = <JobInfo>[];

  for (final entry in jobs.entries) {
    final jobKey = entry.key as String;
    final job = entry.value;
    if (job == null || job is! Map) continue;

    final stepsList = job['steps'];
    if (stepsList == null || stepsList is! List) continue;

    final runsOn = job['runs-on'] is String ? job['runs-on'] as String : null;

    final steps = <JobStep>[];
    for (final step in stepsList) {
      if (step == null || step is! Map) continue;

      final name = (step['name'] as String?) ?? '';

      if (step['uses'] != null) {
        Map<String, String>? withParams;
        if (step['with'] is Map) {
          withParams = {};
          for (final e in (step['with'] as Map).entries) {
            withParams[e.key.toString()] = e.value.toString();
          }
        }
        steps.add(
          JobStep(
            name: name,
            uses: step['uses'] as String,
            withParams: withParams,
          ),
        );
      } else if (step['run'] != null) {
        steps.add(JobStep(name: name, run: step['run'] as String));
      }
    }

    if (steps.isEmpty) continue;

    final needs = <String>[];
    final rawNeeds = job['needs'];
    if (rawNeeds is List) {
      needs.addAll(rawNeeds.cast<String>());
    } else if (rawNeeds is String) {
      needs.add(rawNeeds);
    }

    jobInfos.add(
      JobInfo(jobKey: jobKey, needs: needs, runsOn: runsOn, steps: steps),
    );
  }

  return jobInfos;
}

String workflowFileDocId(
  String teamId,
  String repository,
  String branch,
  String fileName,
) {
  return '${teamId}_${repository.replaceAll('/', '_')}_${branch}_$fileName';
}

File: firebase/ts_functions/src/github/workflowParser.ts
export interface JobStep {
  name: string;
  run?: string;
  uses?: string;
  withParams?: Record<string, string>;
}

export interface JobInfo {
  jobKey: string;
  needs: string[];
  runsOn?: string;
  steps: JobStep[];
}

type JsonMap = Record<string, unknown>;

export function matchesTrigger(
  parsed: JsonMap,
  triggerType: string,
  triggerBranch?: string,
): boolean {
  const on = parsed.on;
  if (on === undefined || on === null) return false;
  const yamlTriggerKey = triggerType === "pullRequest" ? "pull_request" : triggerType;

  if (typeof on === "string") return on === yamlTriggerKey;
  if (Array.isArray(on)) return on.includes(yamlTriggerKey);
  if (typeof on === "object") {
    const triggers = on as JsonMap;
    if (!(yamlTriggerKey in triggers)) return false;
    const triggerConfig = triggers[yamlTriggerKey];
    if (triggerConfig === null || triggerConfig === undefined) return true;
    if (typeof triggerConfig === "object") {
      const branches = (triggerConfig as JsonMap).branches;
      if (branches !== undefined && triggerBranch !== undefined) {
        const branchList = Array.isArray(branches) ? branches.map(String) : [String(branches)];
        if (!branchList.includes(triggerBranch)) return false;
      }
    }
    return true;
  }
  return false;
}

export function extractJobs(parsed: JsonMap): JobInfo[] {
  const jobs = parsed.jobs;
  if (typeof jobs !== "object" || jobs === null || Array.isArray(jobs)) return [];
  const jobInfos: JobInfo[] = [];

  for (const [jobKey, job] of Object.entries(jobs as JsonMap)) {
    if (typeof job !== "object" || job === null || Array.isArray(job)) continue;
    const jobMap = job as JsonMap;
    if (!Array.isArray(jobMap.steps)) continue;
    const steps: JobStep[] = [];
    for (const step of jobMap.steps) {
      if (typeof step !== "object" || step === null || Array.isArray(step)) continue;
      const stepMap = step as JsonMap;
      const name = typeof stepMap.name === "string" ? stepMap.name : "";
      if (typeof stepMap.uses === "string") {
        let withParams: Record<string, string> | undefined;
        if (
          typeof stepMap.with === "object" &&
          stepMap.with !== null &&
          !Array.isArray(stepMap.with)
        ) {
          withParams = Object.fromEntries(
            Object.entries(stepMap.with as JsonMap).map(([key, value]) => [key, String(value)]),
          );
        }
        steps.push({ name, uses: stepMap.uses, ...(withParams ? { withParams } : {}) });
      } else if (typeof stepMap.run === "string") {
        steps.push({ name, run: stepMap.run });
      }
    }
    if (steps.length === 0) continue;
    const needs = Array.isArray(jobMap.needs)
      ? jobMap.needs.map(String)
      : typeof jobMap.needs === "string"
        ? [jobMap.needs]
        : [];
    jobInfos.push({
      jobKey,
      needs,
      runsOn: typeof jobMap["runs-on"] === "string" ? jobMap["runs-on"] : undefined,
      steps,
    });
  }

  return jobInfos;
}

export function workflowFileDocId(
  teamId: string,
  repository: string,
  branch: string,
  fileName: string,
): string {
  return `${teamId}_${repository.replaceAll("/", "_")}_${branch}_${fileName}`;
}
